// @angular-architects/module-federation-runtime/esm2022/enhanced/lib/init-federation.mjs
import { init } from '@module-federation/enhanced/runtime';
let config = {};
export async function initFederation(manifest, options) {
    if (typeof manifest === 'string') {
        config = await loadManifest(manifest);
    }
    else {
        config = parseConfig(manifest);
    }
    const runtimeConfig = toRuntimeConfig(config, options);
    return init(runtimeConfig);
}
export function toRuntimeConfig(config, options) {
    return {
        //
        // The runtime assumes an empty string as the name for
        // the host. Alternatively, we have to pass the same
        // name to withFederation (compile time config) and
        // initFederation (runtime time config on app start)
        //
        name: '',
        ...options?.runtimeOptions,
        remotes: [
            ...(options?.runtimeOptions?.remotes ?? []),
            ...toRemotes(config),
        ],
    };
}
function toRemotes(config) {
    return Object.values(config).map((c) => ({
        name: c.name,
        entry: (c.entry ?? c.remoteEntry ?? ''),
        type: c.type,
    }));
}
export function getManifest() {
    return config;
}
//  Just needed to align with original webpack-based plugin
export async function setManifest(manifest) {
    config = parseConfig(manifest);
}
export async function loadManifest(configFile) {
    const result = await fetch(configFile);
    if (!result.ok) {
        throw Error('could not load configFile: ' + configFile);
    }
    config = parseConfig(await result.json());
    return config;
}
export function parseConfig(config) {
    const result = {};
    for (const key in config) {
        const value = config[key];
        let entry;
        if (typeof value === 'string') {
            entry = {
                name: key,
                remoteEntry: value,
                type: 'module',
            };
        }
        else {
            entry = {
                ...value,
                name: key,
                type: value.type || 'module',
            };
        }
        result[key] = entry;
    }
    return result;
}
