// @angular-architects/module-federation/rspack.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.shareAll = exports.share = exports.findRootTsConfigJson = exports.SharedMappings = exports.DEFAULT_SKIP_LIST = exports.DEFAULT_SECONARIES_SKIP_LIST = void 0;
const tslib_1 = require("tslib");
tslib_1.__exportStar(require("./src/rspack"), exports);
var webpack_1 = require("./webpack");
Object.defineProperty(exports, "DEFAULT_SECONARIES_SKIP_LIST", { enumerable: true, get: function () { return webpack_1.DEFAULT_SECONARIES_SKIP_LIST; } });
Object.defineProperty(exports, "DEFAULT_SKIP_LIST", { enumerable: true, get: function () { return webpack_1.DEFAULT_SKIP_LIST; } });
Object.defineProperty(exports, "SharedMappings", { enumerable: true, get: function () { return webpack_1.SharedMappings; } });
Object.defineProperty(exports, "findRootTsConfigJson", { enumerable: true, get: function () { return webpack_1.findRootTsConfigJson; } });
Object.defineProperty(exports, "share", { enumerable: true, get: function () { return webpack_1.share; } });
Object.defineProperty(exports, "shareAll", { enumerable: true, get: function () { return webpack_1.shareAll; } });
